MM-15635 - Making markdown tables more responsive

avasconcelos114 opened the pull request:
#### Summary
Adds more responsiveness to markdown tables

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/11949

#### Checklist
- [x] Has UI changes

#### Device Information
This PR was tested on: 
* Google Pixel 3 Emulator (Android 9)

#### Screenshots
* Channel view:
![image](https://user-images.githubusercontent.com/29700158/64779969-458c1900-d59a-11e9-8141-e0814dfbcc18.png)

![image](https://user-images.githubusercontent.com/29700158/64780003-550b6200-d59a-11e9-9e20-db94f64f11cf.png)

* Full table view:
  * As flexbox
![image](https://user-images.githubusercontent.com/29700158/64780025-5e94ca00-d59a-11e9-9a8d-3966cdc65b71.png)

  * Fixed cell width (pannable)
![image](https://user-images.githubusercontent.com/29700158/64779780-d31b3900-d599-11e9-8eb1-e42691e0a933.png)

* Landscape
![image](https://user-images.githubusercontent.com/29700158/64779891-11185d00-d59a-11e9-835d-6fae33672102.png)

enahum commented on app/components/markdown/markdown_table/index.js:
```diff
@@ -7,12 +7,15 @@ import {connect} from 'react-redux';
 import {getTheme} from 'mattermost-redux/selectors/entities/preferences';
 
 import {goToScreen} from 'app/actions/navigation';
+import {isTablet, isLandscape} from 'app/selectors/device';
```
there is a more accurate way to detect if the device is a tablet

`IS_TABLET` from DeviceTypes constants.

Also I would try and avoid isLandscape and use `Dimensions` from ReactNative to determine if it is landscape or not.

Let me know if you need more details.

avasconcelos114 replied:
Ah i wasn't aware of these two methods, seems like a simple enough fix :)

Will change it right away

enahum commented on app/components/markdown/markdown_table/markdown_table.js:
```diff
@@ -25,6 +25,8 @@ export default class MarkdownTable extends React.PureComponent {
         numColumns: PropTypes.number.isRequired,
         theme: PropTypes.object.isRequired,
+        isTablet: PropTypes.bool.isRequired,
+        isLandscape: PropTypes.bool.isRequired,
```
see above

esethna commented:
@avasconcelos114 thanks for the PR! @enahum can we build this for PR testing?

mattermod commented:
`Build Apps for PR` label detected. Will generate the app if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Android build App: https://releases.mattermost.com/mattermost-mobile-pr/PR-3249/Mattermost_Beta.apk)

iOS build App: itms-services://?action=download-manifest&url=https://releases.mattermost.com/mattermost-mobile-pr/PR-3249/Mattermost_Beta.plist

esethna commented:
Queued for QA review to test different table sizes, screen sizes and device orientations

hmhealey commented on app/components/markdown/markdown_table/markdown_table.js:
```diff
@@ -67,21 +69,50 @@ export default class MarkdownTable extends React.PureComponent {
     };
 
-    handleContentSizeChange = (contentWidth, contentHeight) => {
+    handleContentSizeChange = (_containerHeight, contentHeight) => {
```
Isn't this first parameter still the width of the content? Also, we don't generally prefix variables with an underscore, so it would be nice if you could change this back

avasconcelos114 replied:
Ah yes, my mistake with calling it `containerHeight` instead, I'll just revert that back to its original value

hmhealey commented on app/components/markdown/markdown_table/markdown_table.js:
```diff
@@ -67,21 +69,50 @@ export default class MarkdownTable extends React.PureComponent {
+
+        const renderAsFlex = this.shouldRenderAsFlex(isFullView);
+        if (renderAsFlex) {
+            tableStyle.push({flex: 1});
```
Can you add an entry to the `StyleSheet` for this?

hmhealey commented on app/components/markdown/markdown_table/markdown_table.js:
```diff
@@ -122,22 +154,22 @@ export default class MarkdownTable extends React.PureComponent {
             <TouchableOpacity onPress={this.handlePress}>
                 <ScrollView
-                    contentContainerStyle={{width: this.getTableWidth()}}
+                    contentContainerStyle={{width: '100%'}}
```
You might be able to remove this prop since I think this is the default value for the prop. If not though, would you mind adding it to the `StyleSheet` as well?

hmhealey commented on app/screens/table/table.js:
```diff
@@ -12,25 +12,30 @@ export default class Table extends React.PureComponent {
     render() {
-        const content = this.props.renderRows();
+        const content = this.props.renderRows(true);
+        const style = this.props.renderAsFlex ? {flex: 1} : {width: this.props.tableWidth};
```
And same comment here

avasconcelos114 commented:
Made a few more changes regarding variable names and adding new styles to the `StyleSheet`  

Though in cases where we get the width value from the `getTableWidth` function I still had to keep it as `{width: this.getTableWidth()}`. If there are any better ways to handle these cases let me know :)

saturninoabril commented:
Thank you Andre for your contribution!

1. Header should have background color
2. 1-column span should take up max column width
![Screen Shot 2019-09-21 at 2 50 37 AM](https://user-images.githubusercontent.com/5334504/65351749-795ee280-dc1b-11e9-9742-f9b77564f24a.png)
![Screen Shot 2019-09-21 at 2 50 23 AM](https://user-images.githubusercontent.com/5334504/65351776-85e33b00-dc1b-11e9-8304-827f0a7b0409.png)

3. 3-column span should not show fade on right side
![Screen Shot 2019-09-21 at 2 53 40 AM](https://user-images.githubusercontent.com/5334504/65351959-fa1dde80-dc1b-11e9-8ad7-90644f9480fc.png)

4. On full screen view, Height should be based on cell height 
![Screen Shot 2019-09-21 at 2 55 25 AM](https://user-images.githubusercontent.com/5334504/65352014-20dc1500-dc1c-11e9-90c9-ad869109a749.png)

5. On full screen view, should respect safe area view on landscape 
![Screen Shot 2019-09-21 at 2 55 58 AM](https://user-images.githubusercontent.com/5334504/65352223-ae1f6980-dc1c-11e9-8bf5-2032f44faf9d.png)

avasconcelos114 commented:
Thank you for the feedback!

I'm working on those but I can't help but feel like there is a bit of conflict from between what's [in the design specs](https://projects.invisionapp.com/share/GSSBM9GXVDT#/screens/366701386) and what you mentioned on item `2`, as the specs say a 1 column table should only expand up to the maximum width of 192dp

Should I just have 1 column tables display as a flexbox anyway? (not a hard fix at all but I just want to be sure it conforms to UX specifications)

avasconcelos114 commented:
Patched up all the other items first and tables should be looking a bit better than before now

![image](https://user-images.githubusercontent.com/29700158/65368859-2e77b600-dc81-11e9-97d6-64258d28ea75.png)

Thank you for testing things out on iOS and finding those issues :)

Let me know how to handle the 1-column tables

saturninoabril commented:
Thanks Andre.  For item 1, that's good point.  I'll defer to @andrewbrown00 on what is the __max column with__ for 1-column span table.

andrewbrown00 commented:
@avasconcelos114 thanks for the contribution here 😄. Here are the specs for a single column:

**Min**: 96pd
**Max**: 192dp

avasconcelos114 commented:
At first I thought it was only the horizontal padding that needed to be changed to 8, but then I realized it was both, and replaced those two to a single `padding: 8` value

Sorry about the commit mess

avasconcelos114 commented:
Resolved conflicts with https://github.com/mattermost/mattermost-mobile/pull/3280 and made sure both functionalities worked properly together (more testing needed to be sure)

<img width="383" alt="image" src="https://user-images.githubusercontent.com/29700158/66456599-05408d80-eaa9-11e9-9063-6fc37df25b4b.png">

saturninoabril commented:
Hello Andre, sorry for the delay as I was out last week.  Looking into this now.

avasconcelos114 commented:
No worries!

Let me know if anything needs to be fixed :+1:

saturninoabril commented:
Hello Andre, please see reviews.

Portrait view:
1. 4-column table in fullscreen view - not enough space on right side nor scrollable
![Screen Shot 2019-10-14 at 4 02 29 PM](https://user-images.githubusercontent.com/5334504/66737032-6f887200-ee9d-11e9-9a13-813d3da450d8.png)
2. 5 or more columns don't have expand button
![Screen Shot 2019-10-14 at 4 05 49 PM](https://user-images.githubusercontent.com/5334504/66737731-5254a300-ee9f-11e9-844a-c4c72b7aced5.png)

Landscape view (not specified in the design but may be improved)
3. Does not fill available space in center view
![Screen Shot 2019-10-14 at 4 29 26 PM](https://user-images.githubusercontent.com/5334504/66738270-9c8a5400-eea0-11e9-9d17-1951f0225921.png)
![Screen Shot 2019-10-14 at 4 30 22 PM](https://user-images.githubusercontent.com/5334504/66738279-a1e79e80-eea0-11e9-9e8d-67e26e6325b5.png)

saturninoabril commented:
Issues on Android:
1. Expand button is not per design.  And when pressed/tapped, the button's ripple effect is observed across the width of center view
![Screen Shot 2019-10-14 at 6 37 27 PM](https://user-images.githubusercontent.com/5334504/66745790-8dac9d00-eeb2-11e9-9372-be2a23ee98bb.png)

2. Cells with different width caused to render unaligned column in fullscreen table view.
![Screen Shot 2019-10-14 at 6 39 01 PM](https://user-images.githubusercontent.com/5334504/66745871-ce0c1b00-eeb2-11e9-992d-93e5b2cd7626.png)

avasconcelos114 commented:
Added a new commit that should have done the following things:
* Fixed issue where a 4-column table renders beyond device width in full table view
* Fixed issue where the expand button isn't rendered correctly in tables with 5+ columns
* Changed rules for how to render tables in landscape view (Would be good to have some insight from UI/UX designers on what's the preferred behavior)
* Separated table views between android/iOS to allow for platform specific styling
* Fixed expand buttons not rendering correctly on android
* Changed display type of expand button, so that the icon can be centrally positioned instead of relying on fixed padding values

Great catches on all these issues btw! :)
I'll be sure to start testing on both iOS and Android before pushing anything new

mattermod commented:
`Build Apps for PR` label detected. Will generate the app if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Build the App failed. See the logs for more information.

mattermod commented:
Build the App failed. See the logs for more information.

mattermod commented:
`Build Apps for PR` label detected. Will generate the app if build succeeds (checks pass and no conflicts with base branch).

enahum commented on app/components/markdown/markdown_table/markdown_table.js:
```diff
@@ -54,7 +54,8 @@ export default class MarkdownTable extends React.PureComponent {
             renderRows: this.renderRows,
-            tableWidth: this.getTableWidth(),
+            tableWidth: this.getTableWidth(true),
+            renderAsFlex: this.shouldRenderAsFlex(true),
```
When rotating the device this prop won't get updated.

avasconcelos114 replied:
I've been taking a closer look at this, and while it's true that this prop won't get updated on device rotation, the behavior of tables (whether to render as flexbox, or calculate a fixed width) isn't affected by whether the user is in landscape mode or not

The landscape bit is (currently) only being used on 4-column tables in the channel view, as they should render as a flexbox in landscape mode but otherwise should have a fixed width when in portrait mode. 

So I'm not so sure if using `onLayout` to run another check on whether or not to render as flex would change much here?

enahum replied:
I think you are right, lets keep it as is and in any case we could look for an alternative later on

enahum commented on app/screens/table/table.android.js:
```diff
@@ -0,0 +1,42 @@
+} from 'react-native';
+import {makeStyleSheetFromTheme} from 'app/utils/theme';
+
+export default class Table extends React.PureComponent {
```
Why two files ? they do not differ much is only the props for the nested ScrollView and the displayFlex style, I think combining this two files into one makes more sense and then use the `Platform` module

avasconcelos114 replied:
Ah yep, sure thing

Will merge those back into a single file with two flex styles

enahum replied:
@avasconcelos114 you can do

```js
const getStyleSheet = makeStyleSheetFromTheme(() => {
    return {
        displayFlex: {
            ...Platform.select({
                android: {
                    flex: 1,
                },
                ios: {
                    flex: 0,
                },
            },
        },
    };
});
```

avasconcelos114 replied:
Oh that's neat!

Thanks for sharing this one, didn't know about this alternative :)

enahum commented on app/screens/table/table.ios.js:
```diff
@@ -0,0 +1,40 @@
+} from 'react-native';
+import {makeStyleSheetFromTheme} from 'app/utils/theme';
+
+export default class Table extends React.PureComponent {
```
see comment for Android

mattermod commented:
Android build App: https://releases.mattermost.com/mattermost-mobile-pr/PR-3249/Mattermost_Beta.apk)

iOS build App: itms-services://?action=download-manifest&url=https://releases.mattermost.com/mattermost-mobile-pr/PR-3249/Mattermost_Beta.plist

avasconcelos114 commented:
Merged back the full table view into a single file, using the method suggested by Elias to use separate styles depending on which platform is being used

avasconcelos114 commented:
Rebased and resolved a small conflict with `package-lock.json`

hmhealey commented:
Btw, @esethna is out this week. He should be back to review this next week.

enahum commented:
@esethna please review

esethna commented:
@enahum when you have a chance can you rebuild with the new signing changes?

avasconcelos114 commented:
Ran another rebase to make sure there won't be conflicts or issues building

avasconcelos114 commented:
Oddly only the iOS app is failing to build, issues seem unrelated to this particular PR but if there's anything I can do to fix this let me know! :)

enahum commented:
I’m looking into it

enahum commented:
Alright, I was able to build it https://community.mattermost.com/core/pl/gwku9f359pdxmqy9zrpktj4y1h

for some reason the only way to build it successfully was to ssh into the box, will look into this further to fix it properly

enahum commented:
@avasconcelos114 this is being cut from the release sadly we did not have the time to properly tested and should be included for 1.28.

Regardless of that can you please merge master so we can merge this as soon as it gets tested

avasconcelos114 commented:
Ah yep, no worries!

Right now I'm merging back and resolving conflicts that happened with the other changes to the markdown table

esethna commented:
Looks good overall, a couple comments:

1. It would be nice to have dynamic sizing on columns if there are 5 or more, since at a fixed width if the columns have very little data there is significant blank space
2. I would expect to be able to scroll a table by dragging anywhere on the full screen table view (even if it's the blank space below the table), right now I can only drag on the table itself. Is there a technical restriction here or is there something we can do to make that a better experience.

@andrewbrown00 can you please re-review this as you worked on the designs.

avasconcelos114 commented:
Thanks for the feedback!

1. There's a bit of a limitation here because the tables don't really know how much content each cell has, and every single cell has to be fixed at the same width values to make it look table-like. But I could switch to using the minimum cell width (96) instead of maximum cell width (192) value if necessary

2. That's a good point, I'm pretty sure I can fix it to make it scrollable through the entire area instead

avasconcelos114 commented:
Made some changes with the latest commit:
1. Made the full table view scrollable through any part of the screen, as opposed to only the table area
2. Resolved conflicts with PR that limits tables to 5 columns visible on channel view

Things that will need to be reviewed/discussed are:
1. I did manage to make the full table view scrollable through the entire area on iOS, but the same method does not apply for Android (some complications happen due to the fact that Android requires two stacked `ScrollView` elements to allow for vertical and horizontal scrolling. 
2. I took the liberty to add some logic that renders the `moreRight` element when columns of a table get cut off, as a sign that if a user taps that table they'll be able to see the remaining columns. 

In the picture here you see a table with 5 columns (above) and one with 6 columns (below)
<img width="736" alt="image" src="https://user-images.githubusercontent.com/29700158/69306681-bff8a980-0c6b-11ea-8ca3-e25337e9e340.png">

Let me know what you guys think!

esethna commented:
Thanks Andre, your points above make sense!

@enahum can you fire another build when you have a chance?

enahum commented:
Hey @avasconcelos114 can you please help resolve the conflicts so we can finally have this beauty merged ? Thanks
